File: app/signing.py
"""
HMAC-SHA256 signing for outgoing webhook deliveries.
 
The signature is sent as an X-Webhook-Signature header so receivers can
verify the payload actually came from this service and wasn't tampered
with in transit. The format mirrors what GitHub and Stripe use:
 
    X-Webhook-Signature: sha256=<hex_digest>
 
Receivers verify it by computing the same HMAC over the raw request body
using their copy of the secret, and comparing with constant-time equality.
"""

from __future__ import annotations

import hashlib
import hmac

def sign(payload: bytes, secret: str) -> str:
    """Returns the signature header value, e.g. 'sha256=abc123...'"""

    digest = hmac.new(secret.encode(), payload, hashlib.sha256).hexdigest()
    return f"sha256={digest}"

def verify(payload: bytes, secret: str, signature: str) -> bool:
    """Constant-time verification"""

    expected = sign(payload, secret)
    return hmac.compare_digest(expected, signature)

File: internal/signing/signing.go
// Package signing provies HMAC-SHA256 signing for outgoing webhook deliveries
//
// X-Webhook-Signature: sha256=<hex_digest>
package signing

import (
	"crypto/hmac"
	"crypto/sha256"
	"fmt"
)

// Returns the signature header value for payload signed with the given secret
func Sign(payload []byte, secret string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write(payload)
	return fmt.Sprintf("sha256=%x", mac.Sum(nil))
}

// Checks a signature using constant-time comparison
func Verify(payload []byte, secret, signature string) bool {
	expected := Sign(payload, secret)
	return hmac.Equal([]byte(expected), []byte(signature))
}